### PR TITLE
validator: implement client stub to interatct with proxy server

### PR DESCRIPTION
**Pull request by disksing**
The client stub will look like a RawKVClient / TxnKVClient, but internally send messages to the proxy server.

Signed-off-by: disksing <i@disksing.com>

### CODE REVIEW

File: stub/rawkv.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// DeleteRange deletes all key-value pairs in a range from TiKV.
+func (c *RawClientStub) DeleteRange(startKey, endKey []byte) error {
+	_, err := c.send(fmt.Sprintf("/rawkv/client/%s/delete-range", c.id), &RawRequest{StartKey: startKey, EndKey: endKey})
+	return err
+}
+
+// Scan queries continuous kv pairs in range [startKey, endKey), up to limit pairs.
+func (c *RawClientStub) Scan(startKey, endKey []byte, limit int) ([][]byte, [][]byte, error) {
```
Comment (rleungx) [inline stub/rawkv.go:105]: Do we also need `ReverseScan`?

Comment (disksing) [reply stub/rawkv.go:105]: Yes. Plan to do it when resolving https://github.com/tikv/client-validator/issues/6

File: stub/msg.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stub
+
+// RawRequest is the structure of a rawkv request that the http proxy accepts.\
```
Comment (rleungx) [inline stub/msg.go]:
```suggestion
// RawRequest is the structure of a rawkv request that the http proxy accepts.
```

File: stub/rawkv.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	id          string
+}
+
+// NewRawClientStub creates a client for rawkv calls.
+func NewRawClientStub(proxyServer string, pdServers []string) (*RawClientStub, error) {
+	client := &RawClientStub{
+		client:      http.Client{Timeout: time.Second * 10},
+		proxyServer: strings.TrimSuffix(proxyServer, "/"),
+	}
+	res, err := client.send("/rawkv/client/new", &RawRequest{PDAddrs: pdServers})
```
Comment (rleungx) [inline stub/rawkv.go:42]: How about using a prefix variable?

File: stub/rawkv.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	if err != nil {
+		return nil, err
+	}
+	client.id = res.ID
+	return client, nil
+}
+
+// Close closes the client and releases resources in proxy server.
+func (c *RawClientStub) Close() error {
+	_, err := c.send(fmt.Sprintf("/rawkv/client/%s/close", c.id), &RawRequest{})
```
Comment (rleungx) [inline stub/rawkv.go:52]: maybe we can directly use `+` here?

Comment (disksing) [reply stub/rawkv.go:52]: I think this way is more readable to understand the format of the URI.

File: stub/rawkv.go (hunk 3 above)
Comment (disksing) [reply stub/rawkv.go:42]: I think this way is more readable to understand the format of the URI.

File: stub/rawkv.go (hunk 4 above)
Comment (rleungx) [reply stub/rawkv.go:52]: `fmt.Sprintf` relies on the reflection which is not elegant I think.

Comment (disksing) [reply stub/rawkv.go:52]: Why reflection is not elegant... For the validator, performance is not important, what matters is correctness and readability.

Comment (rleungx) [reply stub/rawkv.go:52]: ok

**Review by rleungx — APPROVED** [review]

**Pull request merged**